### PR TITLE
feat: deployment resource detector

### PR DESCRIPTION
**Pull request by blumamir**
automatic resource detector for [deployment](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/deployment_environment.md)

I am new to ruby, so please don't hesitate to point me on how to do things correctly if I implemented any anti-pattern.
Also, I am not sure how to write tests for extracting the env from `::Rails::env`. Will appreciate some guidance on this

### CODE REVIEW

**Review by mottibec — APPROVED** [review]

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (arielvalentin) [inline resource_detectors/test/opentelemetry/detectors/deployment_test.rb:19]: There is a test helper [`with_env`](https://github.com/open-telemetry/opentelemetry-ruby/blob/50ce54bd7098c1bc280ec99b486a4f1f51f84ce7/common/lib/opentelemetry/common/test_helpers.rb#L7) that should be available for you to use for test cases.


Here are some examples: https://github.com/open-telemetry/opentelemetry-ruby/search?q=with_env

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (arielvalentin) [inline resource_detectors/test/opentelemetry/detectors/deployment_test.rb:28]: Please consider adding a test for the `Rails.env` case. 

You can use helpers like `stub_const` to simulate the existence of classes:

https://github.com/open-telemetry/opentelemetry-ruby/search?q=stub_const

File: resource_detectors/lib/opentelemetry/resource/detectors/auto_detector.rb
Comment (arielvalentin) [inline resource_detectors/lib/opentelemetry/resource/detectors/auto_detector.rb:16]: @open-telemetry/ruby-maintainers I believe we discussed breaking this gem up into separate gems and removing the auto-detector. Is that correct?

Should that be done before we merge this PR?

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (arielvalentin) [inline resource_detectors/test/opentelemetry/detectors/deployment_test.rb:18]: Please consider adding a test case for when the environment variable is `nil` and the `Rails.env` is undefined

```
irb(main):003:0> ENV['RACK_ENV']
=> nil
````

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (arielvalentin) [inline resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:12]: Personally speaking I think it would be OK to rely on only using environment variables instead of framework specific helpers. 

So in this case I think it would be fine to use something like `ENV['RACK_ENV'] || ENV['RAILS_ENV'] || ENV['APP_ENV']`

File: resource_detectors/lib/opentelemetry/resource/detectors/auto_detector.rb
Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/auto_detector.rb:16]: I believe it can be done after this PR is merged unless it creates issues

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:12]: The issue is that the frameworks have a default environment of "development" in case ENV is missing.
To reflect that in the resource, I will need to write `ENV['RACK_ENV'] || ENV['RAILS_ENV'] || ENV['APP_ENV'] || "development"`, which will work for rails and sinatra but will incorrectly report "development" environment if these frameworks are not being used.

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (blumamir) [reply resource_detectors/test/opentelemetry/detectors/deployment_test.rb:19]: Much shorter and cleaner, thanks for pointing me to this function :)
I refactored the code to use it

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (blumamir) [reply resource_detectors/test/opentelemetry/detectors/deployment_test.rb:18]: I meant for the first test "returns an empty resource" to check this case. Is it the case? or should I refactor it to match your suggestion?

Comment (arielvalentin) [reply resource_detectors/test/opentelemetry/detectors/deployment_test.rb:18]: That is fair, although it would mean when executing the test the environment must not have RACK_ENV set otherwise it would result in the test failing.

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (arielvalentin) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:12]: Totally understand. Will you be using sinatra integration in a separate PR?

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (arielvalentin) [inline resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:19]: For symmetry with `rack_env` 😄 

```suggestion
        def rails_env
```

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (blumamir) [reply resource_detectors/test/opentelemetry/detectors/deployment_test.rb:24]: Added it like this:
```rb
      it 'return rails env when present' do
        allow(::Rails).to receive(:env).and_return('env from rails')
        with_env({}) do
          _(detected_resource).must_be_instance_of(OpenTelemetry::SDK::Resources::Resource)
          _(detected_resource_attributes).must_equal('deployment.environment' => 'env from rails')          
        end
      end
```

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:19]: good point :) renamed it to be consistent

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (blumamir) [reply resource_detectors/test/opentelemetry/detectors/deployment_test.rb:30]: good point, added clearing the env before this test:
```rb
      with_env({}) do
        _(detected_resource).must_be_instance_of(OpenTelemetry::SDK::Resources::Resource)
        _(detected_resource_attributes).must_equal(expected_resource_attributes)
      end
```

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (arielvalentin) [reply resource_detectors/test/opentelemetry/detectors/deployment_test.rb:36]: Looks like this test is failing in CI.

You'll have to stub the `Rails` constant since it doesn't exist. 

```
NameError: uninitialized constant Rails
5431
    /home/runner/work/opentelemetry-ruby/opentelemetry-ruby/resource_detectors/test/opentelemetry/detectors/deployment_test.rb:22:in `block (4 levels) in <top (required)>'
5432
```

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb]: I wasn't planing to add Sinatra support for this PR.
Tried to extract the env like this:
```rb 
          def sinatra_env
            # https://github.com/sinatra/sinatra/blob/e69b6b9dee7165d3a583fc8a6af10ceee1ea687d/lib/sinatra/base.rb#L1801
            ::Sinatra::Base::environment.to_s if defined?(::Sinatra::Base::environment)
          end  
```

Which works fine if the `require 'sinatra'` is placed before the SDK is initialized.
If user configures the SDK and only then require 'sinatra', then this will fail. In addition, it looks like the sinatra instrumentation does not work in this case (we should probably open an issue and debug it as an other PR).

Do you have any suggestion on how to extract the value in this case?

Comment (arielvalentin) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb]: Thank you for giving it a try and thank you for your patience as at this point I've written more comments that actual code in this PR. 

I think it is reasonable to add development dependencies for testing. Require the dependencies as part of the test case and it will make them available to the system under test. You can do this as part of Bundle.require or explicitly requiring it in the test case.

That combined with mocking and `unstub_cost` helpers should be enough to do the trick.

Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb]: Thank you too for the high-quality review and help on these topics.
did some experimenting with sinatra. I was able to get it to work with the help of @YanivD . If you can take another look at the updated code it could be great :)

As documented in the code, this resource detector will not be able to extract the default enviroment "development" if sinatra is used and it is required after resource detector is run (resource is immutable and there is no way to know if it will be rrequired in the future at the time of resource creation).

This is the relevant part of the code for reference:
```rb
        def sinatra_env
          # https://github.com/sinatra/sinatra/blob/e69b6b9dee7165d3a583fc8a6af10ceee1ea687d/lib/sinatra/base.rb#L1801
          # cases:
          #
          # 1. if sinatra is "require"d before the detector, then we return the value from the library
          # this case will return the default "development" fallback if not env variable is set which is good.
          #
          # 2. if sinatra is "require"d after the detector, then:
          # 2.1 if user is setting environment via 'APP_ENV' or 'RACK_ENV' then those value will be picked up and reported
          # 2.2 else, the sinatra environment will fallback to "development", but detector will return nil for it.
          # this issue is not covered, as when detector initialize the immutable resource, it has no way
          # of knowing if "sinatra" will be required later or not.
          (::Sinatra::Base.environment.to_s if defined?(::Sinatra::Base.environment)) || ENV['APP_ENV']
        end
```

File: resource_detectors/test/opentelemetry/detectors/deployment_test.rb
Comment (blumamir) [reply resource_detectors/test/opentelemetry/detectors/deployment_test.rb:65]: done. tests should pass now :)

**Comment by blumamir** [comment]
@arielvalentin , thank you again for the detailed review, appreciate it.
CI is not green and I think I addressed all the comments. Can you please resolve the conversations which are solved and let me know if there is anything else missing?

File: resource_detectors/test/test_helper.rb
Comment (fbogsany) [reply resource_detectors/test/test_helper.rb:13]: We have preferred not to pull in RSpec related dependencies, despite a strong preference for RSpec on the part of some maintainers (_cough_ @robertlaurin).

Comment (arielvalentin) [reply resource_detectors/test/test_helper.rb:13]: hmmm upon further inspection it does not look like the test cases are using `rspec-mocks`. Am I mistaken?

Comment (blumamir) [reply resource_detectors/test/test_helper.rb:13]: I added it so I can use the `stub_const`. copied it from other packages in the repo:
image

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (ericmustin) [inline resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:42]:
```suggestion
          (::Sinatra::Base.environment.to_s if (defined?(::Sinatra) && defined?(::Sinatra::Base))  || ENV['APP_ENV']
```

File: resource_detectors/test/test_helper.rb
Comment (arielvalentin) [reply resource_detectors/test/test_helper.rb:13]: Oh my mistake. I could have sworn that was coming from minitest-stub-const. 😳

Thanks again for your patience

**Review by arielvalentin — APPROVED** [review]

**Review by ericmustin — CHANGES_REQUESTED** [review]
Some small feedback on the implementation itself, but broadly I have some hesitations about this as a resource detector. Let me try to lay these out

1. The name, `Deployment` detector is confusing, i think partly because there's nowhere in our repo to let someone add details about what the detector does. If `resource_detectors` is going to grow, I think adding a section to `resource_detectors/README.md`, that briefly describes this resource detector, and improving the code comments here https://github.com/open-telemetry/opentelemetry-ruby/blob/474236b24db7b8894f0f73ff2ce0abc7e63f963e/resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb#L6 to inform users that that this resource detector specifically enables detection of the `environment` for Rails, Sinatra, and Rack based applications, would go a long way.

2. Is there any prior art that you're basing this approach on? As a pattern for resource detectors, its unusual to me, I generally expect resource detectors to deal with the platform, like gcp/aws/alibaba/etc, and not so much the application libraries. I'm not sure its necessarily bad or not permitted to take the approach being taken here, but it would help my confidence in maintaining this if there were other languages which have taken a similar approach. I briefly reviewed the js resource detectors and couldnt find anything similar to this one.

**Comment by arielvalentin** [comment]
@ericmustin good catches. Those slipped right past me 😓

**Comment by blumamir** [comment]
Hi @ericmustin , Thanks for the review :)

> The name, Deployment detector is confusing

I implemented the resource detector based on [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/deployment_environment.md). I think that as long as it's in the spec, we need to follow the naming and behaviour as described there. Your comment can be a good issue to open in the spec repo.

> i think partly because there's nowhere in our repo to let someone add details about what the detector does

Agree. I was generally following the pattern for the GCP detector in the repo and tried not to introduce any new concepts. Maybe one the detectors are broken down into different gems, we can add a brief details in each gem's README.md, WDYT?

> Is there any prior art that you're basing this approach on? 

I implemented this resource detector for js [here](https://github.com/aspecto-io/opentelemetry-ext-js/tree/master/detectors/node/resource-detector-deployment). This detector is hosted in a separate repo since the native js detectors api is async and we needed a sync detector. It's been included in Aspecto's otel distro for long time and has been working well.

> As a pattern for resource detectors, its unusual to me, I generally expect resource detectors to deal with the platform, like gcp/aws/alibaba/etc, and not so much the application libraries. I'm not sure its necessarily bad or not permitted to take the approach being taken here, but it would help my confidence in maintaining this if there were other languages which have taken a similar approach. I briefly reviewed the js resource detectors and couldnt find anything similar to this one.

I try to look at it from the perspective of the user. When looking at a trace it can be very helpful to know if it's coming from production \ staging \ ci \ tests \ development, as this can aid in filtering relevant traces as well as explaining the behaviour observed in the trace. Unfortunately, the environment configuration in ruby is not integrated into the platform itself (like with node's NODE_ENV variable), but is handled by the frameworks. That is the reason I followed the behavior of the framework to extract this info. I am open to suggestions on alternatives to supplying this resource attribute :)

File: resource_detectors/README.md
Comment (blumamir) [reply resource_detectors/README.md:33]: The resource setter for the configurator is actually [merging the resource itself](https://github.com/open-telemetry/opentelemetry-ruby/blob/a37637c241cacc1799dd0728248c2d1372986348/sdk/lib/opentelemetry/sdk/configurator.rb#L70):

```rb
      # Accepts a resource object that is merged with the default telemetry sdk
      # resource. The use of this method is optional, and is provided as means
      # to include additional resource information.
      # If a resource key collision occurs the passed in resource takes priority.
      #
      # @param [Resource] new_resource The resource to be merged
      def resource=(new_resource)
        @resource = @resource.merge(new_resource)
      end
```

I was confused myself when writing this, until I looked at the source code for that. Maybe we should consider making it clearer in the configurator so it's less confusing?

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:23]: I was actually starting with just checking the `::Rails` and then something crashed in my playground application when verifying that the detector is working.
Will try to reproduce it and paste here an example so we can analyse it. Is it problematic in any way to check for `defined?(::Rails.env)` instead of `defined?(::Rails)`?

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:42]: good catch, for some reason I thought the `defined?` method is able to handle it.
Is there any optional chaining operator similar to `&.` that can work for modules?

File: resource_detectors/README.md
Comment (ericmustin) [reply resource_detectors/README.md:33]: Oh hm, yea i guess it's fine if the configurator handles the merge automagically. Disregard.

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (ericmustin) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:42]: i did some testing and it actually can handle it just fine! TIL, disregard

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (ericmustin) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:23]: after some testing it seems fine to check `::Rails.env`. 

Would be curious to know what the error was. I just modified my suggestion above, i think we want to be invoking `.to_s` and not `.to_str` here, as there is a difference. `.to_str` returns a "string like object" while `.to_s` returns a string representation of the object. 

Frankly I'm not sure if we even need to invoke `.to_s` at all, since afaik `Rails.env` ought to return a string, but if there's any doubt then `.to_s` would be appropriate (and `.to_str` would likely be incorrect)

**Comment by ericmustin** [comment]
@blumamir thanks for answering my questions. I think I'm fine with this, I do think the deeper issue is that these all ought to be split into separate gems, which would allow us to be a little bit more descriptive about what the resource detectors do, and give users more flexibility about which ones they want to use. I have one small feedback item that needs to be addressed that i just added a comment about, and beside that i think this LGTM

**Review by SomalianIvan — CHANGES_REQUESTED** [review]
This seems like it has good motivation, but basing resource detection on the environment variables seems like guessing to me. I believe the detection of the framework specific attributes on the resource should belong to instrumentation. E.g. `rails_env_check` should be part of rails instrumentation.

File: resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb
Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb]: > i think we want to be invoking .to_s and not .to_str here, as there is a difference. .to_str returns a "string like object" while .to_s returns a string representation of the object.

Thanks for spotting that, I wasn't aware of the difference, was just trying `to_str` and it worked. Changed to use `to_s`

> Frankly I'm not sure if we even need to invoke .to_s at all, since afaik Rails.env ought to return a string, but if there's any doubt then .to_s would be appropriate (and .to_str would likely be incorrect)

The value that is [returned from rails env function is: `ActiveSupport::EnvironmentInquirer`](https://github.com/rails/rails/blob/5647a9c1ced68d20338552d47a3b755e10a271c4/railties/lib/rails.rb#L74).
From the docs:

```
  # Wrapping a string in this class gives you a prettier way to test
  # for equality. The value returned by <tt>Rails.env</tt> is wrapped
  # in a StringInquirer object, so instead of calling this:
  #
  #   Rails.env == 'production'
  #
  # you can call this:
  #
  #   Rails.env.production?
  #
```

This is the reason `to_s` is mandatory for the env value before we store it into the resource

Comment (blumamir) [reply resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb]: > Would be curious to know what the error was

So I wasn't able to reproduce it in a real setup, it works good with both `defined?(::Rails.env)` and `defined?(::Rails)`, and also for sinatra.
But if I remove the `env` from the check, current test suite fails with things like:
```
  1) Error:
OpenTelemetry::Resource::Detectors::Deployment::.detect#test_0001_returns an empty resource:
NoMethodError: undefined method `env' for Rails:Module
Did you mean?  end
    /Users/amirblum/repos/opentelemetry-ruby/resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:23:in `rails_env'
    /Users/amirblum/repos/opentelemetry-ruby/resource_detectors/lib/opentelemetry/resource/detectors/deployment.rb:12:in `detect'
    /Users/amirblum/repos/opentelemetry-ruby/resource_detectors/test/opentelemetry/detectors/deployment_test.rb:9:in `block (3 levels) in <top (required)>'
```

Any thought on this? should I update the tests somehow or just leave it the way it is?

**Comment by blumamir** [comment]
> This seems like it has good motivation, but basing resource detection on the environment variables seems like guessing to me. I believe the detection of the framework specific attributes on the resource should belong to instrumentation. E.g. `rails_env_check` should be part of rails instrumentation.

If we detect the environment in the framework instrumentation, then it will be set as span attributes only for this framework spans. I think it's also a valid strategy but has the following cons:
1. only framework spans will be marked with the environment, which means the user will not be able to tell the environment based on downstream service span or trace which started not by the framework.
2. it's less performant, as a resource is immutable for the service and is reported only once per batch, and possibly stored more efficiently in backends. If we publish it as span attribute, it will have to be reported again and again for each framework span.
3. The environment is a value that is static to the run of the service, thus I believe the most suitable place for it is in the resource.
4. Since the Deployment resource detector is defined in the specification, there is a value IMO in supplying such detector for ruby as well. This will give users a unified experience between SDKs in different languages, and backend processing pipelines can rely on common behavior to ingest trace data coming from different sources (they can be written to look for the "deployment.environment" in resource)


> but basing resource detection on the environment variables seems like guessing to me

I agree that this implementation is not 100% robust (as I stated in the comments), and there are edge cases in which it will not work, but IMO, this is the best we can do as long as the resource is an immutable entity in OTEL (which is a problem for many other use-cases as well).

Maybe we can supply a configuration option to additionally capture environment in relevant instrumentations, and then user can choose between installing a resource detector or enabling this option in instrumentation?

**Review by YanivD — APPROVED** [review]

**Comment by blumamir** [comment]
As discussed in the SIG - this detector probably belongs to a future contrib repo, since:
1. It only applies to rails, sinatra and rack, and will not report env for other frameworks
2. If multiple frameworks are used, then only one arbitrary value will be used
3. Fetching values from environment variables will not always reflect the real environment set by the framework/tools/ conventions
4. The name "Deployment" can be a bit confusing as the only attribute detected is "environment".

So I'm closing the PR for this repo and will add it in the future to some other repo that better fits it's characters.

Thanks all for the reviews and the discussion

**Pull request closed without merging**
